### PR TITLE
Simplification of dispatch logic/definition of new observables

### PR DESCRIPTION
**Pull request by PhilipVinc**
Our funny @alleSini99 recently contributed a set of Renyi entropy estimators, which are defined to inherit from `ÀbstractOperator`, so we need to define some methods like `ìs_hermitian` that do not make much sense for such object. 

Moreover, to define the gradient, the dispatch rule for this observable has this ugly-as-hell `TrueT`or `Literal[True]` that nobody besides me understands.

This PR is an attempt to
 - Simplify the creation of a new generic operator/observable
 - Simplify the definition of signatures for dispatch of expect/grad by:
   - remove `use_covariance` argument from the general interface
   - only keep `use covariance` for the expectation value of operators where it make sense, and it will not be part of the dispatch signature

In practice...

- This introduces a new super type of AbstractOperator which I call `AbstractObservable`. The difference between Abstract Operator and AbstractObservable is that an Observable is very general and requires nothing besides an Hilbert space. No is hermitian or dtype arguments. So it should cover the most general case.

- Renyi entropy estimator is transitioned to this interface.

- The signature that users must define for expectation value estimators will now be
```python
@dispatch
def expect(vs: MCState, ob: Reny2Entropy, chunk_size: Optional[int]):
  pass
```
and for gradients will be (the much simpler)
```python
@dispatch
def expect_and_grad(vs: MCState, ob: Reny2Entropy, chunk_size: Optional[int]):
  pass
```

Incidentally, this will make it simple to implement different types of chunking like @chrisrothUT wants to do in #1590 by dispatching on a tuple of integers for the chunk size instead of just an integer. Right now the dispatch logic is very messy and this would not be easy to do.

Note that users are required to specify the chunk size, and if thy don-t support it they have to explicitly state `chunk_size: None`. I could relax this requirement but it makes user-defined code future-proofed in case we add more arguments.

The main problem with those changes is that it breaks user-defined operators using the past syntax.  This is not strictly a problem because this part of the interface is documented to be unstable, though it's annoying. 
I could add some inspect magic to detect usages of the old signatures and auto-convert them to the new format and warn. To be experimented with.

### CODE REVIEW

File: netket/operator/_abstract_observable.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+import abc
+
+from netket.hilbert import AbstractHilbert
+
+
+class AbstractObservable(abc.ABC):
+    """Abstract class for quantum Observables.
+    This class prototypes the methods
+    needed by a class satisfying the Operator concept.
```
Comment (gcarleo) [inline netket/operator/_abstract_observable.py:23]: why the operator concept?

File: netket/operator/_abstract_observable.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    """
+
+    _hilbert: AbstractHilbert
+    r"""The hilbert space associated to this operator."""
```
Comment (gcarleo) [inline netket/operator/_abstract_observable.py:27]: operator?

File: netket/vqs/full_summ/expect.py [hunk 3]
```diff
@@ -65,12 +65,39 @@ def expect(vstate: FullSumState, Ô: DiscreteOperator) -> Stats:  # noqa: F811
...
+        )
+
+
+@expect_and_forces.dispatch
+def expect_and_forces_fullsum(
     vstate: FullSumState,
     Ô: DiscreteOperator,
     *,
-    mutable: CollectionFilter,
+    mutable: CollectionFilter = False,
```
Comment (gcarleo) [inline netket/vqs/full_summ/expect.py:100]: is CollectionFilter documented somewhere?

File: netket/operator/_abstract_observable.py (hunk 1 above)
Comment (PhilipVinc) [reply netket/operator/_abstract_observable.py]: yeah, fixed and expanded.

File: netket/operator/_abstract_observable.py (hunk 2 above)
Comment (PhilipVinc) [reply netket/operator/_abstract_observable.py]: thanks

File: netket/vqs/full_summ/expect.py (hunk 3 above)
Comment (PhilipVinc) [reply netket/vqs/full_summ/expect.py:100]: We don't need this annotation in truth. Dian had added it a while back but we could also remove it. It's a flax concept. In the documentation it should link to flax...

File: netket/operator/_abstract_observable.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    An observable is a general object that defines a quantity that
+    can be comptued starting from a variational state. Observables
+    should be computed using the method `expect` of the variational
+    states, and derivatives of this expectation value can sometimes
+    also be computed using the method `expect_and_grad`.
+
+    All operators are Observables, but some observables are not
+    operators (for example, the entanglement entropy observable
+    does is not an operator).
+
```
Comment (alleSini99) [inline netket/operator/_abstract_observable.py:32]: Typo "the entanglement entropy observable does is not an operator" -> "the entanglement entropy observable is not an operator"

**Review by alleSini99 — COMMENTED** [review]
Fine

**Comment by PhilipVinc** [comment]
I updated the documentation for those changes.

If there are no contrary people, I will merge

**Comment by gcarleo** [comment]
All checks are failing?

File: docs/advanced/custom_operators.ipynb [hunk 5]
```diff
@@ -6,7 +6,7 @@
    "id": "44701f85-3e69-4e1c-b017-54a843297568",
```
Comment (gcarleo) [inline docs/advanced/custom_operators.ipynb:6]: <p>now I am confused, why this is deriving from AbstractObservable and not AbstractOperator? (the text still says AbstractOperator btw) </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/netket/netket/pull/1605/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

File: netket/operator/_abstract_observable.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+
+from netket.hilbert import AbstractHilbert
+
+
+class AbstractObservable(abc.ABC):
```
Comment (gcarleo) [inline netket/operator/_abstract_observable.py:20]: this concept is very useful indeed and more general than AbstractOperator. However, the current abstract interface does not do much, it just contains hilbert, so I am a bit uncertain whether it warrants creating an abstract class? aren't there other methods that should be part of the abstract interface of an Observable/ that are more characteristic than just containing a Hilbert?

**Pull request merged**